mantidproject/mantid
# Move sum banks step and fix bug in ISIS Reflectometry workflow

**Description of work**

**Purpose of work**

To ensure that the sum banks step is in the right place in the ISIS Reflectometry reduction workflow. Also only perform summing when a detector region of interest has been passed in to ensure that other instruments do not inadvertently pick up the summing step.



**Summary of work**

The sum banks step has been move into `ReflectometryReductionOneAuto` so that it happens after the flood correction.

I have fixed a bug in the logic that determines when to sum banks so that this only happens when the data in the workspace is exclusively from the 2D detector. After discussing with our scientists, I have also added the condition that summing will only be considered if a detector ROI has been selected. This is important because the summing step is specific to the new INTER 2D detector,  it is not necessarily relevant for all 2D detectors. This meant I needed to slightly change the workflow of the preview tab so that we don't automatically sum all banks on load. Instead the reduction is now attempted without any summing and the user can then start selecting detector banks and ROI/processing instructions to improve the result. Note that with the `INTER45455` 2D test dataset the preview reduction gives an error until either the detector banks have been summed or an appropriate set of processing instructions have been entered, but this accurately mirrors what happens if the reduction is run normally (via the Runs tab). At some point we will be updating the default processing instructions in the INTER parameter file to something that should prevent this error in normal use.

I have added tests for the changes to `ReflectometryReductionOneAuto3` via additional tests in `ReflectometryISISLoadAndProcessTest`, as the sum banks algorithm is written in Python and calling to Python algorithms from the C++ unit tests doesn't seem to work. This fix is a priority and we expect to make changes to the sum banks step in the future, so at that point it may either no longer be necessary to add tests to `ReflectometryReductionOneAuto3Test` or we can resolve the problem at that time.

Similarly I'm not convinced by the solution for hiding the summed workspaces in this PR, but a better approach can be found as part of future work.

**To test:**


1) Load an INTER flood workspace file from the archive into the ADS - ask Rachel for help with finding this if needed.
1) Open the ISIS Reflectometry interface
1) On the Runs tab, enter `INTER45455` into the first row of the table and angle `2.3`.
1) In the Experiment Settings tab, ensure the Flood Correction drop-down is set to `Workspace`. In the Flood Workspace drop-down, select the workspace you loaded in step 1. In the Lookup Settings table, enter `2001-2240` into the ROIDetectorIDs column.
1) Back on the Runs tab, click the process button. The reduction should complete successfully. There should be a workspace called `TOF_45455_summed_segment` in the `TOF` workspace group. There should also be a call to `ReflectometryReductionOneAuto -> ReflectometryISISSumBanks` in the history of the `IvsQ_binned_45455` workspace, which should be after the call to `ApplyFloodWorkspace`.
1) Clear the ADS and close and re-open the GUI to start again.
1) In the Experiment Settings tab of the GUI, in the Lookup Settings table enter `6-62` in the ROI column (note, this is a different column to the ROIDetectorIDs column we set previously, which should be left blank).
1) On the Runs tab of the GUI, enter `INTER45455` into the first row of the table and angle `2.3`.
1) Click the process button. The reduction should complete successfully, but this time there should be no `TOF_45455_summed_segment` in the `TOF` workspace group and no call to `ReflectometryISISSumBanks` in the history.
1) Finally check running some reductions in the preview tab.

Fixes #35629. 






---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
